components/shared/base/generic_channel.rs

```diff
@@ -30,20 +38,93 @@ pub enum GenericSender<T: Serialize> {
 
 impl<T: Serialize> Serialize for GenericSender<T> {
     fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
-        match self {
-            GenericSender::Ipc(i) => i.serialize(s),
-            GenericSender::Crossbeam(_) => panic!("{GENERIC_CHANNEL_USAGE_ERROR_PANIC_MSG}"),
+        use serde::ser::Error;
+        match &self.0 {
+            GenericSenderVariants::Ipc(sender) => {
+                s.serialize_newtype_variant("GenericSender", 0, "Ipc", sender)
+            },
+            // All GenericSenders will be IPC channels in multi-process mode, so sending a
+            // GenericChannel over existing IPC channels is no problem and won't fail.
+            // In single-process mode, we can also send GenericSenders over other GenericSenders
+            // just fine, since no serialization is required.
+            // The only reason we need / want serialization is to support sending GenericSenders
+            // over existing ipc_channels **in single process mode**. This allows us to
+            // incrementally port channels to the GenericChannel, without needing to follow a
+            // Top-to-bottom approach.
+            // Long-term we can remove this branch in the code again and replace it with
+            // unreachable, since likely all IPC-channels would be GenericChannels.
+            GenericSenderVariants::Crossbeam(sender) => {
+                if opts::get().multiprocess {
+                    return Err(S::Error::custom(
+                        "Crossbeam channel found in multiprocess mode!",
+                    ));
+                } // We know everything is in one address-space, so we can "serialize" the sender by
+                // sending a leaked Box pointer.
+                let sender_clone_addr = Box::leak(Box::new(sender.clone())) as *mut _ as usize;
+                s.serialize_newtype_variant("GenericSender", 1, "Crossbeam", &sender_clone_addr)
+            },
         }
     }
 }
 
-impl<'a, T: Serialize> Deserialize<'a> for GenericSender<T> {
+struct GenericSenderVisitor<T> {
+    marker: PhantomData<T>,
+}
+
+impl<'de, T: Serialize + Deserialize<'de>> serde::de::Visitor<'de> for GenericSenderVisitor<T> {
+    type Value = GenericSender<T>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("enum GenericSender")
```

Comment:
The [doc](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#required-methods) suggests that it should complete the sentence "This Visitor expects to receive..."

```suggestion
        formatter.write_str("a GenericSender variant'")
```